File: Capitulo3_VRCardboard - JAVA/app/src/main/java/com/upv/pm_2022/sep_dic/capitulo3_vrcardboard/NfcSensor.java
package com.upv.pm_2022.sep_dic.capitulo3_vrcardboard;

import android.content.*;
import android.os.*;
import java.io.*;
import android.net.*;
import android.nfc.*;
import android.nfc.tech.*;
import android.util.*;
import android.app.*;
import java.util.*;

public class NfcSensor {
    private static NfcSensor sInstance;
    private final Context mContext;
    private final NfcAdapter mNfcAdapter;
    private final Object mTagLock;
    private final List<ListenerHelper> mListeners;
    private IntentFilter[] mNfcIntentFilters;
    private Ndef mCurrentNdef;
    private Tag mCurrentTag;
    private boolean mCurrentTagIsCardboard;
    private Timer mNfcDisconnectTimer;
    private int mTagConnectionFailures;
    
    public static NfcSensor getInstance(final Context context) {
        if (NfcSensor.sInstance == null) {
            NfcSensor.sInstance = new NfcSensor(context);
        }
        return NfcSensor.sInstance;
    }
    
    private NfcSensor(final Context context) {
        super();
        this.mContext = context.getApplicationContext();
        this.mNfcAdapter = NfcAdapter.getDefaultAdapter(this.mContext);
        this.mListeners = new ArrayList<ListenerHelper>();
        this.mTagLock = new Object();
        if (this.mNfcAdapter == null) {
            return;
        }
        final IntentFilter ndefIntentFilter = new IntentFilter("android.nfc.action.NDEF_DISCOVERED");
        ndefIntentFilter.addAction("android.nfc.action.TECH_DISCOVERED");
        ndefIntentFilter.addAction("android.nfc.action.TAG_DISCOVERED");
        this.mNfcIntentFilters = new IntentFilter[] { ndefIntentFilter };
        this.mContext.registerReceiver((BroadcastReceiver)new BroadcastReceiver() {
            public void onReceive(final Context context, final Intent intent) {
                NfcSensor.this.onNfcIntent(intent);
            }
        }, ndefIntentFilter);
    }
    
    public void addOnCardboardNfcListener(final OnCardboardNfcListener listener) {
        if (listener == null) {
            return;
        }
        synchronized (this.mListeners) {
            for (final ListenerHelper helper : this.mListeners) {
                if (helper.getListener() == listener) {
                    return;
                }
            }
            this.mListeners.add(new ListenerHelper(listener, new Handler()));
        }
    }
    
    public void removeOnCardboardNfcListener(final OnCardboardNfcListener listener) {
        if (listener == null) {
            return;
        }
        synchronized (this.mListeners) {
            for (final ListenerHelper helper : this.mListeners) {
                if (helper.getListener() == listener) {
                    this.mListeners.remove(helper);
                }
            }
        }
    }
    
    public boolean isNfcSupported() {
        return this.mNfcAdapter != null;
    }
    
    public boolean isNfcEnabled() {
        return this.isNfcSupported() && this.mNfcAdapter.isEnabled();
    }
    
    public boolean isDeviceInCardboard() {
        synchronized (this.mTagLock) {
            return this.mCurrentTagIsCardboard;
        }
    }
    
    public NdefMessage getTagContents() {
        synchronized (this.mTagLock) {
            return (this.mCurrentNdef != null) ? this.mCurrentNdef.getCachedNdefMessage() : null;
        }
    }

    public void onResume(final Activity activity) {
        if (!this.isNfcEnabled()) {
            return;
        }
        final Intent intent = new Intent("android.nfc.action.NDEF_DISCOVERED");
        intent.setPackage(activity.getPackageName());
        final PendingIntent pendingIntent = PendingIntent.getBroadcast(this.mContext, 0, intent, 0);
        this.mNfcAdapter.enableForegroundDispatch(activity, pendingIntent, this.mNfcIntentFilters, (String[][])null);
    }
    
    public void onPause(final Activity activity) {
        if (!this.isNfcEnabled()) {
            return;
        }
        this.mNfcAdapter.disableForegroundDispatch(activity);
    }
    
    public void onNfcIntent(final Intent intent) {
        if (!this.isNfcEnabled() || intent == null || !this.mNfcIntentFilters[0].matchAction(intent.getAction())) {
            return;
        }
        this.onNewNfcTag((Tag)intent.getParcelableExtra("android.nfc.extra.TAG"));
    }
    
    private void onNewNfcTag(final Tag nfcTag) {
        if (nfcTag == null) {
            return;
        }
        synchronized (this.mTagLock) {
            final Tag previousTag = this.mCurrentTag;
            final Ndef previousNdef = this.mCurrentNdef;
            final boolean previousTagWasCardboard = this.mCurrentTagIsCardboard;
            this.closeCurrentNfcTag();
            this.mCurrentTag = nfcTag;
            this.mCurrentNdef = Ndef.get(nfcTag);
            if (this.mCurrentNdef == null) {
                if (previousTagWasCardboard) {
                    this.sendDisconnectionEvent();
                }
                return;
            }
            boolean isSameTag = false;
            if (previousNdef != null) {
                final byte[] tagId1 = this.mCurrentTag.getId();
                final byte[] tagId2 = previousTag.getId();
                isSameTag = (tagId1 != null && tagId2 != null && Arrays.equals(tagId1, tagId2));
                if (!isSameTag && previousTagWasCardboard) {
                    this.sendDisconnectionEvent();
                }
            }
            NdefMessage nfcTagContents;
            try {
                this.mCurrentNdef.connect();
                nfcTagContents = this.mCurrentNdef.getCachedNdefMessage();
            }
            catch (Exception e) {
                final String s = "NfcSensor";
                final String s2 = "Error reading NFC tag: ";
                final String value = String.valueOf(e.toString());
                Log.e(s, (value.length() != 0) ? s2.concat(value) : new String(s2));
                if (isSameTag && previousTagWasCardboard) {
                    this.sendDisconnectionEvent();
                }
                return;
            }
            this.mCurrentTagIsCardboard = this.isCardboardNdefMessage(nfcTagContents);
            if (!isSameTag && this.mCurrentTagIsCardboard) {
                synchronized (this.mListeners) {
                    for (final ListenerHelper listener : this.mListeners) {
                        listener.onInsertedIntoCardboard(CardboardDeviceParams.createFromNfcContents(nfcTagContents));
                    }
                }
            }
            if (this.mCurrentTagIsCardboard) {
                this.mTagConnectionFailures = 0;
                (this.mNfcDisconnectTimer = new Timer("NFC disconnect timer")).schedule(new TimerTask() {
                    @Override
                    public void run() {
                        synchronized (NfcSensor.this.mTagLock) {
                            if (!NfcSensor.this.mCurrentNdef.isConnected()) {
                                ++NfcSensor.this.mTagConnectionFailures;
                                if (NfcSensor.this.mTagConnectionFailures > 1) {
                                    NfcSensor.this.closeCurrentNfcTag();
                                    NfcSensor.this.sendDisconnectionEvent();
                                }
                            }
                        }
                    }
                }, 250L, 250L);
            }
        }
    }
    
    private void closeCurrentNfcTag() {
        if (this.mNfcDisconnectTimer != null) {
            this.mNfcDisconnectTimer.cancel();
        }
        if (this.mCurrentNdef == null) {
            return;
        }
        try {
            this.mCurrentNdef.close();
        }
        catch (IOException e) {
            Log.w("NfcSensor", e.toString());
        }
        this.mCurrentTag = null;
        this.mCurrentNdef = null;
        this.mCurrentTagIsCardboard = false;
    }
    
    private void sendDisconnectionEvent() {
        synchronized (this.mListeners) {
            for (final ListenerHelper listener : this.mListeners) {
                listener.onRemovedFromCardboard();
            }
        }
    }
    
    private boolean isCardboardNdefMessage(final NdefMessage message) {
        if (message == null) {
            return false;
        }
        for (final NdefRecord record : message.getRecords()) {
            if (this.isCardboardNdefRecord(record)) {
                return true;
            }
        }
        return false;
    }
    
    private boolean isCardboardNdefRecord(final NdefRecord record) {
        if (record == null) {
            return false;
        }
        final Uri uri = record.toUri();
        return uri != null && CardboardDeviceParams.isCardboardUri(uri);
    }
    
    private static class ListenerHelper implements OnCardboardNfcListener
    {
        private OnCardboardNfcListener mListener;
        private Handler mHandler;
        
        public ListenerHelper(final OnCardboardNfcListener listener, final Handler handler) {
            super();
            this.mListener = listener;
            this.mHandler = handler;
        }
        
        public OnCardboardNfcListener getListener() {
            return this.mListener;
        }
        
        @Override
        public void onInsertedIntoCardboard(final CardboardDeviceParams deviceParams) {
            this.mHandler.post((Runnable)new Runnable() {
                @Override
                public void run() {
                    ListenerHelper.this.mListener.onInsertedIntoCardboard(deviceParams);
                }
            });
        }
        
        @Override
        public void onRemovedFromCardboard() {
            this.mHandler.post((Runnable)new Runnable() {
                @Override
                public void run() {
                    ListenerHelper.this.mListener.onRemovedFromCardboard();
                }
            });
        }
    }
    
    public interface OnCardboardNfcListener
    {
        void onInsertedIntoCardboard(CardboardDeviceParams p0);
        
        void onRemovedFromCardboard();
    }
}

File: VRCardboardKotlin/app/src/main/java/com/vrcardboardkotlin/NfcSensor.kt
package com.vrcardboardkotlin

import android.app.Activity
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.Uri
import android.nfc.NdefMessage
import android.nfc.NdefRecord
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.Ndef
import android.os.Handler
import android.os.Parcelable
import android.util.Log
import java.io.IOException
import java.util.Timer
import java.util.TimerTask

class NfcSensor private constructor(context: Context) {
    private val mContext: Context
    private val mNfcAdapter: NfcAdapter?
    private val mTagLock: Any
    private val mListeners: MutableList<ListenerHelper>
    private val mNfcIntentFilters: Array<IntentFilter>
    private var mCurrentNdef: Ndef? = null
    private var mCurrentTag: Tag? = null
    private var mCurrentTagIsCardboard: Boolean = false
    private var mNfcDisconnectTimer: Timer? = null
    private var mTagConnectionFailures: Int = 0

    init {
        this.mContext = context.applicationContext
        this.mNfcAdapter = NfcAdapter.getDefaultAdapter(this.mContext)
        this.mListeners = ArrayList()
        this.mTagLock = Any()

        if (this.mNfcAdapter != null) { // Solo ejecuta el código si mNfcAdapter no es null
            val ndefIntentFilter = IntentFilter("android.nfc.action.NDEF_DISCOVERED").apply {
                addAction("android.nfc.action.TECH_DISCOVERED")
                addAction("android.nfc.action.TAG_DISCOVERED")
            }
            this.mNfcIntentFilters = arrayOf(ndefIntentFilter)

            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
                mContext.registerReceiver(
                    object : BroadcastReceiver() {
                        override fun onReceive(context: Context, intent: Intent) {
                            this@NfcSensor.onNfcIntent(intent)
                        }
                    },
                    ndefIntentFilter,
                    Context.RECEIVER_NOT_EXPORTED // Asegura que el receptor no sea accesible fuera de tu app
                )
            } else {
                mContext.registerReceiver(object : BroadcastReceiver() {
                    override fun onReceive(context: Context, intent: Intent) {
                        this@NfcSensor.onNfcIntent(intent)
                    }
                }, ndefIntentFilter)
            }
        } else {
            this.mNfcIntentFilters = arrayOf() // Asigna un valor vacío para evitar problemas
        }
    }



    fun addOnCardboardNfcListener(listener: OnCardboardNfcListener?) {
        if (listener == null) {
            return
        }
        synchronized(this.mListeners, {
            for (helper: ListenerHelper in this.mListeners) {
                if (helper.getListener() === listener) {
                    return
                }
            }
            mListeners.add(ListenerHelper(listener, Handler()))
        })
    }

    fun removeOnCardboardNfcListener(listener: OnCardboardNfcListener?) {
        if (listener == null) {
            return
        }
        synchronized(this.mListeners, {
            for (helper: ListenerHelper in this.mListeners) {
                if (helper.getListener() === listener) {
                    mListeners.remove(helper)
                }
            }
        })
    }

    val isNfcSupported: Boolean
        get() {
            return this.mNfcAdapter != null
        }

    val isNfcEnabled: Boolean
        get() {
            return this.isNfcSupported && mNfcAdapter!!.isEnabled()
        }

    val isDeviceInCardboard: Boolean
        get() {
            synchronized(this.mTagLock, {
                return this.mCurrentTagIsCardboard
            })
        }

    val tagContents: NdefMessage?
        get() {
            synchronized(this.mTagLock, {
                return if ((this.mCurrentNdef != null)) mCurrentNdef!!.getCachedNdefMessage() else null
            })
        }

    fun onResume(activity: Activity) {
        if (!this.isNfcEnabled) {
            return
        }
        val intent: Intent = Intent("android.nfc.action.NDEF_DISCOVERED")
        intent.setPackage(activity.getPackageName())
        val pendingIntent: PendingIntent = PendingIntent.getBroadcast(this.mContext, 0, intent, 0)
        mNfcAdapter!!.enableForegroundDispatch(
            activity,
            pendingIntent,
            this.mNfcIntentFilters,
            null as Array<Array<String?>?>?
        )
    }

    fun onPause(activity: Activity?) {
        if (!this.isNfcEnabled) {
            return
        }
        mNfcAdapter!!.disableForegroundDispatch(activity)
    }

    fun onNfcIntent(intent: Intent?) {
        if (!this.isNfcEnabled || (intent == null) || !mNfcIntentFilters.get(0)
                .matchAction(intent.getAction())
        ) {
            return
        }
        this.onNewNfcTag(intent.getParcelableExtra<Parcelable>("android.nfc.extra.TAG") as Tag?)
    }

    private fun onNewNfcTag(nfcTag: Tag?) {
        if (nfcTag == null) {
            return
        }
        synchronized(this.mTagLock, {
            val previousTag: Tag? = this.mCurrentTag
            val previousNdef: Ndef? = this.mCurrentNdef
            val previousTagWasCardboard: Boolean = this.mCurrentTagIsCardboard
            this.closeCurrentNfcTag()
            this.mCurrentTag = nfcTag
            this.mCurrentNdef = Ndef.get(nfcTag)
            if (this.mCurrentNdef == null) {
                if (previousTagWasCardboard) {
                    this.sendDisconnectionEvent()
                }
                return
            }
            var isSameTag: Boolean = false
            if (previousNdef != null) {
                val tagId1: ByteArray? = mCurrentTag!!.getId()
                val tagId2: ByteArray? = previousTag!!.getId()
                isSameTag = ((tagId1 != null) && (tagId2 != null) && tagId1.contentEquals(tagId2))
                if (!isSameTag && previousTagWasCardboard) {
                    this.sendDisconnectionEvent()
                }
            }
            val nfcTagContents: NdefMessage
            try {
                mCurrentNdef!!.connect()
                nfcTagContents = mCurrentNdef!!.getCachedNdefMessage()
            } catch (e: Exception) {
                val s: String = "NfcSensor"
                val s2: String = "Error reading NFC tag: "
                val value: String = e.toString().toString()
                Log.e(
                    s,
                    if ((value.length != 0)) (s2 + value) else (java.lang.String(s2) as String)
                )
                if (isSameTag && previousTagWasCardboard) {
                    this.sendDisconnectionEvent()
                }
                return
            }
            this.mCurrentTagIsCardboard = this.isCardboardNdefMessage(nfcTagContents)
            if (!isSameTag && this.mCurrentTagIsCardboard) {
                synchronized(this.mListeners, {
                    for (listener: ListenerHelper in this.mListeners) {
                        listener.onInsertedIntoCardboard(
                            CardboardDeviceParams.Companion.createFromNfcContents(
                                nfcTagContents
                            )
                        )
                    }
                })
            }
            if (this.mCurrentTagIsCardboard) {
                this.mTagConnectionFailures = 0
                (Timer("NFC disconnect timer").also({ this.mNfcDisconnectTimer = it })).schedule(
                    object : TimerTask() {
                        override fun run() {
                            synchronized(this@NfcSensor.mTagLock, {
                                if (!mCurrentNdef!!.isConnected()) {
                                    ++this@NfcSensor.mTagConnectionFailures
                                    if (this@NfcSensor.mTagConnectionFailures > 1) {
                                        this@NfcSensor.closeCurrentNfcTag()
                                        this@NfcSensor.sendDisconnectionEvent()
                                    }
                                }
                            })
                        }
                    },
                    250L,
                    250L
                )
            }
        })
    }

    private fun closeCurrentNfcTag() {
        if (this.mNfcDisconnectTimer != null) {
            mNfcDisconnectTimer!!.cancel()
        }
        if (this.mCurrentNdef == null) {
            return
        }
        try {
            mCurrentNdef!!.close()
        } catch (e: IOException) {
            Log.w("NfcSensor", e.toString())
        }
        this.mCurrentTag = null
        this.mCurrentNdef = null
        this.mCurrentTagIsCardboard = false
    }

    private fun sendDisconnectionEvent() {
        synchronized(this.mListeners, {
            for (listener: ListenerHelper in this.mListeners) {
                listener.onRemovedFromCardboard()
            }
        })
    }

    private fun isCardboardNdefMessage(message: NdefMessage?): Boolean {
        if (message == null) {
            return false
        }
        for (record: NdefRecord in message.getRecords()) {
            if (this.isCardboardNdefRecord(record)) {
                return true
            }
        }
        return false
    }

    private fun isCardboardNdefRecord(record: NdefRecord?): Boolean {
        if (record == null) {
            return false
        }
        val uri: Uri? = record.toUri()
        return uri != null && CardboardDeviceParams.Companion.isCardboardUri(uri)
    }

    private class ListenerHelper(
        private val listener: OnCardboardNfcListener,
        private val mHandler: Handler
    ) : OnCardboardNfcListener {

        fun getListener(): OnCardboardNfcListener {
            return listener
        }

        override fun onInsertedIntoCardboard(deviceParams: CardboardDeviceParams?) {
            mHandler.post {
                listener.onInsertedIntoCardboard(deviceParams)
            }
        }

        override fun onRemovedFromCardboard() {
            mHandler.post {
                listener.onRemovedFromCardboard()
            }
        }
    }


    interface OnCardboardNfcListener {
        fun onInsertedIntoCardboard(p0: CardboardDeviceParams?)

        fun onRemovedFromCardboard()
    }

    companion object {
        private var sInstance: NfcSensor? = null
        fun getInstance(context: Context): NfcSensor? {
            if (sInstance == null) {
                sInstance = NfcSensor(context)
            }
            return sInstance
        }
    }
}
